src/app/generate/generate.go

```diff
@@ -196,7 +196,8 @@ func tagVersionGetter(cCtx *cli.Context) (*git.TagsSource, error) {
 		tagOpts = append(tagOpts, git.TagsMatching("^"+matching))
 	}
 
-	src, err := git.NewRepoTagsGetter(cCtx.String(gitRootFlag), tagOpts...)
+	workDir := cCtx.String(gitRootFlag)
+	src, err := git.NewRepoTagsGetter(workDir, git.NewRepoCommitsGetter(workDir), tagOpts...)
```

Comment:
I'm not a fan of putting a `RepoCommitsGetter` inside the `RepoTagsGetter`, I don' think the TagsGetter needs Commits to perform its basic functionality.

If I understood the code correctly, what we want to do here is to _filter_ tags after we have collected them, right? Meaning, we don't use any information from the tags that is not returned as `[]Tag` by `RepoTagsGetter.Tags`.

If this is the case I think that what we should do here is have a TagsFilterer that can work on what `RepoTagsGetter` returns.

```go
tagGetter := git.NewRepoTagsGetter(workDir, tagOpts...) // Same signature, object, and behavior as before this PR
tags := tagGetter.Tags()
//...
filterer := ReachableTagsFilterer(git.NewRepoCommitsGetter(workDir))
tags = filterer.Filter(tags)
```

Reply:
I assumed that the basic functionality was retrieving the tags from a specific branch.

However, I can extract it, I just fear that in the future someone could use `.Tag`  without knowing that the filter is not applied by default

Reply:
would make sense to extract as well:
```
match         *regexp.Regexp
[...]
func TagsMatching(regex string) TagOptionFunc
```

Reply:
Now it follows the approach of the tags regex and it is "pluggable". 
If we want a separate all filtering then I would create something like `NewCommitFilter` for both the regex and the branch

Reply:
> I see it as a similar functionality to: `func TagsMatching(regex string) TagOptionFunc`

This is a good point. The biggest difference I see is that matching a regex is a very easy thing to do (we don't require any dependency in the object), but filtering from a list of commits it's not, as it requires you to inject the CommitsGetter in the equation.

Adding `func TagsMatchingCommits(source CommitSource) TagOptionFunc` looks fine to me.

Extracting both regex filtering _and_ commit filtering out looks even better. In the end, we also have a tag replacement function, that is already extracted and performed by the `TagsSource`:
https://github.com/newrelic/release-toolkit/blob/e6cbc226f6c96c11d389232cfa7f92ce6774437e/src/git/tag_source.go#L35-L35

Reply:
Ok, then we can leave it like this and wait for a major refactor when the v1 comes